AOT for Fuchsia.

rmacnak-google opened the pull request:
The main complication here is the ordering of initialization. On iOS and Android, the AOT snapshot comes from a fixed location on the filesystem, so the engine has been able to initialize the VM early. In Fuchsia, we don't have access to an AOT snapshot until we receive the first app bundle, so we have to delay initializing the VM until then. The AOT snapshots should be refactored such that the VM can be initialize and the service isolate started independently of any app isolate.

abarth commented on content_handler/BUILD.gn:
```diff
@@ -54,7 +54,83 @@ executable("content_handler") {
     "//flutter/glue",
     "//flutter/lib/ui",
-    "//flutter/runtime",
+    "//flutter/runtime:runtime",
```
the ":runtime" here is redundant.

abarth commented on content_handler/BUILD.gn:
```diff
@@ -54,7 +54,83 @@ executable("content_handler") {
+    "//apps/mozart/services/buffers/cpp",
+    "//apps/mozart/services/composition",
+    "//apps/mozart/services/input",
+    "//apps/mozart/services/views",
```
Can we use gn variables to avoid repeating ourselves so much?

abarth commented on content_handler/runtime_holder.cc:
```diff
@@ -103,6 +111,57 @@ void RuntimeHolder::Init(
+
+  // TODO(rmacnak): We should generate the AOT vm snapshot separately from
+  // each app so we can initialize before receiving the first app bundle.
+  static std::atomic_bool first_app(true);
```
There no need for atomic.  There's only one thread here.

chinmaygarde replied:
Even if there were many threads, wouldn't you need to do a compare and exchange?

abarth commented on runtime/BUILD.gn:
```diff
@@ -109,3 +109,68 @@ source_set("runtime") {
+    "test_font_data.cc",
+    "test_font_data.h",
+    "test_font_selector.cc",
+    "test_font_selector.h",
```
Same comment about using variables to avoid repeating ourselves.

abarth commented:
![LGTM](http://i3.cpcache.com/product/1351564863/lgtm_bumper_stickers.jpg?width=200&height=200)

rmacnak-google commented:
Refactoring the GN revealed the AOT content handler was linking with both libdart_precompiled_runtime and lib_dart (the JIT), and the ordering of the deps affects which one wins. I'll untangle this knot before landing this CL.

rmacnak-google commented:
Please take a look:
 - Refactored GN to avoid duplication and remove use of libdart in favor of an explicit libdart_jit or libdart_precompiled_runtime
 - Added missing language flags to the gen_snapshot invocation.
